samples/cmake/tls_test/main.cpp

```diff
@@ -0,0 +1,48 @@
+#include <coreinit/thread.h>
+#include <coreinit/time.h>
+
+#include <whb/proc.h>
+#include <whb/log.h>
+#include <whb/log_console.h>
+
+#include <thread>
+
+__thread
+__attribute((tls_model("global-dynamic")))
```

Comment:
Is this required even with `-ftls-model=global-dynamic` ?

Most C++ code will just be using [thread_local](https://en.cppreference.com/w/cpp/keyword/thread_local) so it will be a shame if we cannot support unmodified code.

Reply:
sadly yes, I couldn't get `gcc` to use compatible relocations, even when compiling with `-ftls-model=global-dynamic`. it seems to ignore that flag unless specified directly as a variable attribute.

even when using `-msdata=eabi`, which clearely states that `R2` is reservered for `sdata2_base`, gcc will just keep emitting code using `R_PPC_TP*` instead of `R_PPC_DTP*` relocations, expecting `R2` to hold the Thread Pointer for whatever reason.

the only solution right now is to just `#define` `thread_local` as `thead_local __attribute((tls_model("global-dynamic")))`. it might even work as a compile flag.

Reply:
```
tls_model ("tls_model")
The tls_model attribute sets thread-local storage model (see Thread-Local) of a particular __thread variable, overriding -ftls-model= command-line switch on a per-variable basis. The tls_model argument should be one of global-dynamic, local-dynamic, initial-exec or local-exec.

Not all targets support this attribute.
```

I suspect the `-ftls-model=global-dynamic` flag might not be passed properly to the compiler. Can you port this example over to standard Makefiles and see if it works that way?

Reply:
I did test it with a standard markefile and the result was the same. the flag does get passed, gcc just decided that it knows better and optimized it away.
from https://gcc.gnu.org/onlinedocs/gcc/Code-Gen-Options.html
>Alter the thread-local storage model to be used (see Thread-Local). The model argument should be one of ‘global-dynamic’, ‘local-dynamic’, ‘initial-exec’ or ‘local-exec’. Note that the choice is subject to optimization: the compiler may use a more efficient model for symbols not visible outside of the translation unit, or if -fpic is not given on the command line.

Reply:
Alright, nice find. At this point I'm wondering if this requires us to patch the compiler to avoid this "optimization" - I believe that would be more preferable to needing to work around the TLS model every single time a TLS variable is used.

Reply:
that would certainly be preferable, but it might be a bit harder than it looks.

the main problem here seems to be a bug in `gcc` that makes it emit code that contradicts the chosen ABI. as long as it thinks that `R2` has a thread pointer, it will be very hard to convince it to drop the optimizations.

who knows, maybe it is even possible to make `gcc` restrict itself to `R_PPC_DTPMOD32` and `R_PPC_DTPREL32`, which would make patching relocations in `elf2rpl` unnecessary.